### DISCUSSION TITLE
Can we customize BAR 1 size via linux kernel/module params?

### DISCUSSION THREAD

**Discussion by Qubitium** [NVIDIA/open-gpu-kernel-modules#579] [Q&A]
Right now I am testing A100 80GB PCIE on a consumer board and it is allocating 128GB BAR 1 size. Unfortunately this is causing pcie issues where other gpu devices are unable to allocate BAR due to the A100 hogging all the available space. 

I have searched everywhere but unable to find a solution to have the A100 allocate only 16GB of BAR 1 for example. Via lspci output, the A100 supports 16GB and lower BAR 1 size via power of 2 increments. 

Is a way to lower the A100 or any other Nvidia GPU to a lower but supported BAR 1 size? 

Thank you for any assistance.

**Comment by aaronp24** [comment]
You might be able to use the `NVreg_EnableResizableBar` feature but you'll need to modify the code if you want to specifically request a smaller size: https://github.com/NVIDIA/open-gpu-kernel-modules/blob/main/kernel-open/nvidia/nv-pci.c#L170

That code always tries the largest size first and goes down from there if it doesn't fit. You'll need to disable some of the bits in the `sizes` variable in order to resize it smaller. (Note: I haven't tried this)

**Reply by Qubitium** [reply]
@aaronp24  This worked perfectly. Unfortunately, it did not fix my core issue. Very strange. Maybe I can pick you brain for a bit. 

Before
-------
```
1. A100 GPU assigned 128GB BAR 1 (it is at max 128GB bar limit)
2. 2X4090 GPU on same pcie root, assigned 256MB BAR 1 (it is capable of 32GB bar via lspci)
```

So I though decreasing the A100 bar size at kernel/driver load will solve this, using your tip. 

After compiling custom nvidia driver and making sure A100 only assigned 32GB. 
```
1. A100, assigned 32GB BAR1 (success!)
2. 2x4090, on same pcie root, still assigned 256MB (?)
```

So then I tried using linux kernel (6.6.5) sysfs method to resize bar of the 4090 to 1GB. I figured, I just freed 90GB of bar/mem space from pcie table there should be enough for tiny 1GB for the 2x4090 right?  Wrong, I still get "no space left on device" when attempting to modify bar1 size via sysfs. 

Pcie/bar seems like voodoo magic to me.  Any possible reason for the rejection of 2x 1GB bar1 for the 4090s after 90GB freed  for A100?

**Reply by Qubitium** [reply]
After reducing A100 bar1 to 32GB, I also tried using sysfs to remove the pcie 4090 devices and pcie rescan to add them back to retrigger bar1 increase, still same 256MB limit.

**Reply by Qubitium** [reply]
I can freely use sysfs to runtime resize A100 bar1 within all the supported ranges. Yet 4090 only allows 256MB and fails with no space left on device for all other supported sizes.

**Reply by aaronp24** [reply]
What does `pci_rebar_get_possible_sizes` report for your 4090s? ReBAR could be disabled in the VBIOS, although I thought it was enabled for all 40 series cards.

Oh, it might also be worth checking your PCIe topology. I think if both 4090s are on the same PCIe bridge, it might not be able to move the BARs if it can't also move the memory window of the upstream bridges due to the other GPU still being present. I'm not sure how the upstream kernel handles that kind of situation.

**Comment by atedjarati** [comment]
@Qubitium I am running into similar issues on my PC. I am running aarch64 and need to reduce the BAR 1 size on my graphics card. Can you give me your instructions on how you used sysfs to do this? Thank you!!

**Reply by dagbdagb** [reply]
As root, do:
`find /sys -name resource\?_resize\*`

Find the file representing your device and the BAR in question.

`echo 13 > $file`

The number represents a bit position. 8 is 256MB, 10 is 1GB, 13 is 8GB. etc. 
Upsizing only works if there is free space behind the current allocation.

**Reply by Qubitium** [reply]
What I did was same as @dagbdagb except I unload the driver first or you may get device is in-use error :

Do below as root
```
# resource = your gpu id = find /sys -name resource\?_resize\*

#unload nvidia
modprobe -r nvidia-drm
modprobe -r nvidia-uvm
modprobe -r nvidia

# do resize of BAR1
echo 13 > $resource

# rescan pcie bus
echo 1 > /sys/bus/pci/rescan

#load nvidia
modprobe nvidia
modprobe nvidia-uvm

**Reply by web3creator** [reply]
@Qubitium Hello, how can I customize bar1 to 64gb?

**Reply by web3creator** [reply]
@Qubitium I want to set bar1 of 4090 to 64gb, can you help me?

**Reply by Qubitium** [reply]
@web3creator Why would you want to do this? Are you able to pcb hack 4090 to 64GB? 

4090 has 24GB max. BAR1 size appears to be power of 2 so it's 16GB, 32GB  (24GB 4090) and 64, 128, etc. Having BAR1 that exceeds physical vram doesn't help imho.

**Reply by web3creator** [reply]
@Qubitium 64GB is very helpful for optimizing my program. I can't find any tool to set the size of bar1 to 64GB. The current maximum is 32GB.

**Reply by Qubitium** [reply]
@web3creator  See my above and use 14 for 64GB.   You may need to hack the driver code and up the max ranges. It may only give you the power of 2 closes to your gpu vram. So 24GB gpu is max 13. 

```
# do resize of BAR1
echo 14> $resource
```

**Reply by web3creator** [reply]
@Qubitium  

<img width="793" alt="截屏2025-02-02 20 52 26" src="https://github.com/user-attachments/assets/cf368b79-3963-47e1-a2da-d31946c7da10" />

I execute this and get nothing

**Reply by web3creator** [reply]
> @web3creator See my above and use 14 for 64GB. You may need to hack the driver code and up the max ranges. It may only give you the power of 2 closes to your gpu vram. So 24GB gpu is max 13.
> 
> ```
> # do resize of BAR1
> echo 14> $resource
> ```
hack the driver I don't know how to do it

**Reply by strawberryfly** [reply]
Hey @Qubitium @dagbdagb  thanks for the instruction. When I run echo 16 > [path to resource1_resize], I get the error "Device or resource busy", even though i disabled the nvidia driver. Do you have any idea what might be wrong?

**Reply by dagbdagb** [reply]
You probably need to quote the entire file path or escape any ':'-characters. Let us see your exact command.

**Reply by strawberryfly** [reply]
Here a my commands:
First:
```
find /sys -name resource\?_resize\*

/sys/devices/pci0000:80/0000:80:03.0/0000:84:00.0/resource0_resize
/sys/devices/pci0000:80/0000:80:03.0/0000:84:00.0/resource1_resize
/sys/devices/pci0000:80/0000:80:03.0/0000:84:00.0/resource3_resize
```

second
```
modprobe -r nvidia-drm
modprobe -r nvidia-uvm
modprobe -r nvidia
```

third:
```
echo 16 > '/sys/devices/pci0000:80/0000:80:03.0/0000:84:00.0/resource1_resize'
-bash: echo: write error: No such file or directory
```

```
echo 16 > /sys/devices/pci0000\:80/0000\:80\:03.0/0000\:84\:00.0/resource1_resize 
-bash: echo: write error: No such file or directory
```


```
cd /sys/devices/pci0000\:80/0000\:80\:03.0/0000\:84\:00.0/
echo 16 > resource1_resize
-bash: echo: write error: No such file or directory
```

```
cat resource1_resize
000000000003ffc0

ls -l /sys/devices/pci0000:80/0000:80:03.0/0000:84:00.0/resource1_resize 
-rw-r--r-- 1 root root 4096 Aug  5 06:41 /sys/devices/pci0000:80/0000:80:03.0/0000:84:00.0/resource1_resize
```

**Reply by strawberryfly** [reply]
My device does not officially support resizable bars. I guess that's why it doesn't work :(

**Comment by krpr** [comment]
how to make set a bar size 64GB on a 4090?
lspci report the max option for 4090 is 32GB, is there any way to set it to 64GB or larger?
or this is a limit from vbios that cannot set on linux?

thanks

**Reply by dagbdagb** [reply]
Do you happen to have a 48GB 4090? Not obvious to me there are other scenarios where this makes sense. I do not know the answer to your question, but I find it plausible this is a vbios limitation.

**Reply by krpr** [reply]
yes, a 48gb mod
it should be a vbios limitation
